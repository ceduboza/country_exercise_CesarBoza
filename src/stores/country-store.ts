import { defineStore } from 'pinia';
import type { ICountryUI } from 'src/modules/Countries/domain/entities/ICountryUI';

interface State {
  countries: ICountryUI[],
  countryDetail: ICountryUI | null,
  loading: boolean,
}

export const useCountryStore = defineStore('country', {
  state: (): State => ({
    countries: [],
    countryDetail: null,
    loading: false,
  }),
  getters: {
    isMobile() {
      return document.querySelector('body.mobile') !== null;
    }
  },
  actions: {
    getCountryList() {
      this.loading = true;

      this.$countryDataProvider.getCountries().then((countries) => {
        this.countries = countries;
      }).catch((error) => {
        this.countries = [];
        console.error('Failed to fetch countries:', error);
      }).finally(() => {
        this.loading = false;
      })
    },
    getCountryListByName(name: string) {
      this.loading = true;

      this.$countryDataProvider.getCountriesByName(name).then((countries) => {
        this.countries = countries;
      }).catch((error) => {
        this.countries = [];
        console.error('Failed to fetch countries by name:', error);
      }).finally(() => {
        this.loading = false;
      })
    },
    getCountryDetail(name: string) {
      this.loading = true;
      if (this.countries.length) {
        const country = this.countries.find((country) => country.name.toLowerCase() === name.toLowerCase());
        this.countryDetail = country || null;
        this.loading = false;
      } else {
        this.$countryDataProvider.getCountriesByName(name).then((countries) => {
          const selectedCountry = countries?.[0];
          if (selectedCountry) {
            this.countryDetail = selectedCountry;
          }
        }).catch((error) => {
          console.error('Failed to fetch countries by name:', error);
          this.countryDetail = null;
        }).finally(() => {
          this.loading = false;
        })
      }
    },
  },
});
