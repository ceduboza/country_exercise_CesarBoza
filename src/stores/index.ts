import { defineStore } from '#q-app/wrappers'
import { createPinia } from 'pinia'
import { CountryDataProvider } from 'src/modules/Countries/infraestructure/CountryDataProvider';

export default defineStore(() => {
  const pinia = createPinia()

  pinia.use(({ store }) => {
    store.$countryDataProvider = new CountryDataProvider()
  });

  return pinia;
})