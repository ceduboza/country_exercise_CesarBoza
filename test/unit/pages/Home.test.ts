import { nextTick } from "vue";
import { describe, it, expect, vi } from "vitest";
import { shallowMount } from "@vue/test-utils";
import HomePage from "../../../src/pages/Home/index.vue";
import { useCountryStore } from '../../../src/stores/country-store';
import { QPage } from "quasar";

vi.mock('../../../src/stores/country-store', () => ({
    useCountryStore: vi.fn()
}))

describe("Home", () => {
    let wrapper

    const getCountryList = vi.fn()

    const findQpage = wrapper => wrapper.findComponent({ name: 'Qpage' })
    const findHeaderPage = wrapper => wrapper.findComponent({ name: 'HeaderPage' })
    const findCountryGrid = wrapper => wrapper.findComponent({ name: 'CountryGrid' })
    const findQpDrawer = wrapper => wrapper.findComponent({ name: 'QDrawer' })
    const findCountryFilterForm = wrapper => wrapper.findComponent({ name: 'CountryFilterForm' })
    const findWishlistIcon = wrapper => wrapper.findComponent({ name: 'WishlistIcon' })
    const filterButton = wrapper => wrapper.findComponent({ name: 'QBtn' })

    const contextFactory = () => {
        vi.mocked(useCountryStore).mockReturnValue({
            countries: [],
            loading: false,
            getCountryList,
        } as any)
    }

    const factory = () => {

        contextFactory()

        wrapper = shallowMount(HomePage, {
            global: {
                stubs: {
                    QPage: {
                        name: 'Qpage',
                        template: '<div unit-test-id="q-page"><slot /></div>',
                    },
                    QDrawer: {
                        name: 'QDrawer',
                        template: '<div unit-test-id="q-drawer"><slot /></div>',
                    },
                    CountryGrid: {
                        name: 'CountryGrid',
                        template: `<div unit-test-id="country-grid"><slot name="actions" :country='{ "name": "Test Country" }' /></div>`,
                    },
                    HeaderPage: {
                        name: 'HeaderPage',
                        template: '<div unit-test-id="header-page"><slot /></div>',
                    },
                    QItemLabel: true,
                    QSeparator: true,
                    QBtn: true,
                    CountryFilterForm: true,
                    WishlistIcon: true,
                },
            },
        });
    }

    describe("render", () => {
        it("should render default the component", () => {
            factory()

            expect(wrapper.html()).toMatchSnapshot()
            expect(findQpage(wrapper).exists()).toBe(true)
            expect(findHeaderPage(wrapper).exists()).toBe(true)
            expect(findCountryGrid(wrapper).exists()).toBe(true)
            expect(findQpDrawer(wrapper).exists()).toBe(true)
            expect(findCountryFilterForm(wrapper).exists()).toBe(true)
            expect(findWishlistIcon(wrapper).exists()).toBe(true)
        })
    })
    describe("events", () => {
        it("should call getCountryList when the component is mounted", () => {
            factory()

            expect(getCountryList).toHaveBeenCalled()
        })
        it("should change leftDrawerOpen value when the filter button is clicked", async () => {
            factory()

            const button = filterButton(wrapper)
            button.vm.$emit('click')

            await nextTick()

            const drawer = findQpDrawer(wrapper)

            expect(wrapper.html()).toMatchSnapshot()
            expect(drawer.attributes('modelvalue')).toBe('true')
        })
        it("should change leftDrawerOpen value when countryFilterForm emits onSubmit", async () => {
            factory()

            const button = filterButton(wrapper)
            button.vm.$emit('click')

            await nextTick()

            const countryFilterForm = findCountryFilterForm(wrapper)
            countryFilterForm.vm.$emit('onSubmit')

            await nextTick()

            const drawer = findQpDrawer(wrapper)

            expect(wrapper.html()).toMatchSnapshot()
            expect(drawer.attributes('modelvalue')).toBe('false')
        })
        it("should change leftDrawerOpen value when countryFilterForm emits onReset", async () => {
            factory()

            const button = filterButton(wrapper)
            button.vm.$emit('click')

            await nextTick()

            const countryFilterForm = findCountryFilterForm(wrapper)
            countryFilterForm.vm.$emit('onSubmit')

            await nextTick()

            const drawer = findQpDrawer(wrapper)

            expect(wrapper.html()).toMatchSnapshot()
            expect(drawer.attributes('modelvalue')).toBe('false')
        })
    });
});
