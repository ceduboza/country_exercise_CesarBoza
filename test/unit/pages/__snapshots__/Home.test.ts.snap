// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home > events > should change leftDrawerOpen value when countryFilterForm emits onReset 1`] = `
"<div unit-test-id="q-page" class="contry-grid">
  <div unit-test-id="header-page" title="Countries">
    <q-btn-stub icon="tune" replace="false" disable="false" type="button" flat="false" outline="false" push="false" unelevated="false" square="false" rounded="false" glossy="false" fab="false" fabmini="false" nocaps="false" nowrap="false" dense="false" ripple="true" align="center" stack="false" stretch="false" round="true" darkpercentage="false" class="contry-grid__filter-btn" aria-label="Filters"></q-btn-stub>
  </div>
  <div unit-test-id="country-grid" countries="" isloading="false">
    <wishlist-icon-stub country="[object Object]"></wishlist-icon-stub>
  </div>
  <div unit-test-id="q-drawer" width="300" overlay="true" class="contry-grid__filter-drawer" modelvalue="false">
    <q-item-label-stub overline="false" caption="false" header="true"></q-item-label-stub>
    <q-separator-stub spaced="false" inset="true" vertical="false" light=""></q-separator-stub>
    <country-filter-form-stub></country-filter-form-stub>
  </div>
</div>"
`;

exports[`Home > events > should change leftDrawerOpen value when countryFilterForm emits onSubmit 1`] = `
"<div unit-test-id="q-page" class="contry-grid">
  <div unit-test-id="header-page" title="Countries">
    <q-btn-stub icon="tune" replace="false" disable="false" type="button" flat="false" outline="false" push="false" unelevated="false" square="false" rounded="false" glossy="false" fab="false" fabmini="false" nocaps="false" nowrap="false" dense="false" ripple="true" align="center" stack="false" stretch="false" round="true" darkpercentage="false" class="contry-grid__filter-btn" aria-label="Filters"></q-btn-stub>
  </div>
  <div unit-test-id="country-grid" countries="" isloading="false">
    <wishlist-icon-stub country="[object Object]"></wishlist-icon-stub>
  </div>
  <div unit-test-id="q-drawer" width="300" overlay="true" class="contry-grid__filter-drawer" modelvalue="false">
    <q-item-label-stub overline="false" caption="false" header="true"></q-item-label-stub>
    <q-separator-stub spaced="false" inset="true" vertical="false" light=""></q-separator-stub>
    <country-filter-form-stub></country-filter-form-stub>
  </div>
</div>"
`;

exports[`Home > events > should change leftDrawerOpen value when the filter button is clicked 1`] = `
"<div unit-test-id="q-page" class="contry-grid">
  <div unit-test-id="header-page" title="Countries">
    <q-btn-stub icon="tune" replace="false" disable="false" type="button" flat="false" outline="false" push="false" unelevated="false" square="false" rounded="false" glossy="false" fab="false" fabmini="false" nocaps="false" nowrap="false" dense="false" ripple="true" align="center" stack="false" stretch="false" round="true" darkpercentage="false" class="contry-grid__filter-btn" aria-label="Filters"></q-btn-stub>
  </div>
  <div unit-test-id="country-grid" countries="" isloading="false">
    <wishlist-icon-stub country="[object Object]"></wishlist-icon-stub>
  </div>
  <div unit-test-id="q-drawer" width="300" overlay="true" class="contry-grid__filter-drawer" modelvalue="true">
    <q-item-label-stub overline="false" caption="false" header="true"></q-item-label-stub>
    <q-separator-stub spaced="false" inset="true" vertical="false" light=""></q-separator-stub>
    <country-filter-form-stub></country-filter-form-stub>
  </div>
</div>"
`;

exports[`Home > render > should render default the component 1`] = `
"<div unit-test-id="q-page" class="contry-grid">
  <div unit-test-id="header-page" title="Countries">
    <q-btn-stub icon="tune" replace="false" disable="false" type="button" flat="false" outline="false" push="false" unelevated="false" square="false" rounded="false" glossy="false" fab="false" fabmini="false" nocaps="false" nowrap="false" dense="false" ripple="true" align="center" stack="false" stretch="false" round="true" darkpercentage="false" class="contry-grid__filter-btn" aria-label="Filters"></q-btn-stub>
  </div>
  <div unit-test-id="country-grid" countries="" isloading="false">
    <wishlist-icon-stub country="[object Object]"></wishlist-icon-stub>
  </div>
  <div unit-test-id="q-drawer" width="300" overlay="true" class="contry-grid__filter-drawer" modelvalue="false">
    <q-item-label-stub overline="false" caption="false" header="true"></q-item-label-stub>
    <q-separator-stub spaced="false" inset="true" vertical="false" light=""></q-separator-stub>
    <country-filter-form-stub></country-filter-form-stub>
  </div>
</div>"
`;
